web/modules/custom/nidirect_prisons/src/Controller/WorldpayNotificationController.php

```diff
@@ -0,0 +1,268 @@
+   * succeeded or failed.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The Request.
+   * @return \Symfony\Component\HttpFoundation\Response
+   *   The Response.
+   */
+  public function handleNotification(Request $request) {
+
+    /* See https://docs.worldpay.com/apis/wpg/manage for information
+     * on configuring Worldpay order notifications.
+     *
+     * When a payment request is initialised, the payment transaction is
+     * recorded as 'pending' in the prisoner_payment_transactions table.
+     * When the hosted payment page is submitted, Worldpay sends
+     * notifications on the status of the payment.
+     *
+     * At present, notifications for the following statuses are
+     * configured:
+     *   - 'AUTHORISED'
+     *   - 'CANCELLED'
+     *   - 'SHOPPER_CANCELLED'
+     *   - 'REFUSED'
+     *   - 'ERROR'
+     *
+     * If the payment is AUTHORISED, the payment transaction is
+     * updated to 'success' in the prisoner_payment_transactions table
+     * and in practice this means that the payment has been
+     * successfully approved by the card issuer or bank, but the funds
+     * have not yet been captured or settled (debited from visitor's
+     * account).
+     */
+
+    // Check request is from a Worldpay IP by performing forward and
+    // reverse DNS lookups. Deny access to non-Worldpay IP addresses.
+    $ip = $request->getClientIp();
+
+    // Reverse DNS lookup.
+    $hostname = gethostbyaddr($ip);
+    if (!$hostname || !str_ends_with($hostname, '.worldpay.com')) {
+      \Drupal::logger('nidirect_prisons')->warning('Unrecognized hostname: @hostname from IP @ip', [
+        '@hostname' => $hostname,
+        '@ip' => $ip,
+      ]);
+      return new Response('Access Denied', 403);
+    }
+
+    // Forward DNS lookup.
+    $resolved_ips = gethostbynamel($hostname);
+    if (!$resolved_ips || !in_array($ip, $resolved_ips, TRUE)) {
+      \Drupal::logger('nidirect_prisons')->warning('DNS verification failed for IP @ip with hostname @hostname.', [
+        '@ip' => $ip,
+        '@hostname' => $hostname,
+        '@resolved_ips' => implode(', ', $resolved_ips),
+      ]);
+      return new Response('Access Denied', 403);
+    }
+
+    // Get the raw XML from the request body.
+    $xml_data = $request->getContent();
+
+    // Return 400 bad request if there is something wrong with the
+    // XML. Note Worldpay will retry sending notifications (for up to 7
+    // days) if the response is anything other than a 200 OK.
+
+    if (empty($xml_data)) {
+      \Drupal::logger('nidirect_prisons')->error('Empty Worldpay notification received.');
+      return new Response('Bad request', 400);
+    }
+
+    $xml = simplexml_load_string($xml_data);
+    if (!$xml) {
+      \Drupal::logger('nidirect_prisons')->error('Invalid XML received in Worldpay notification.');
+      return new Response('Bad request', 400);
+    }
+
+    if (!isset($xml->notify)) {
```

Comment:
A future improvement would be to verify the XML against a document schema. https://www.php.net/manual/en/domdocument.schemavalidate.php